Description: Fix ntpviz file encodings
 This fixes ntpviz on python3.  We need to specify the file encoding.
Forwarded: https://gitlab.com/NTPsec/ntpsec/merge_requests/573
Applied-Upstream: https://gitlab.com/NTPsec/ntpsec/commit/03176993a0465257270108e21516ff30aa600f51
Origin: vendor
Author: Richard Laager <rlaager@wiktel.com>
Last-Update: 2017-11-11
--- a/ntpclients/ntpviz
+++ b/ntpclients/ntpviz
@@ -56,6 +56,12 @@
 """)
     sys.exit(1)
 
+if sys.version_info[0] == 2:
+    import codecs
+    def open(file, mode='r', buffering=-1, encoding=None, errors=None):
+        return codecs.open(filename=file, mode=mode, encoding=encoding,
+            errors=errors, buffering=buffering)
+
 # believe it or not, Python has no way to make a simple constant!
 MS_PER_S = 1e3          # milliseconds per second
 NS_PER_S = 1e9          # nanoseconds per second
@@ -379,13 +385,17 @@
     if outfile is None:
         out = None
     else:
-        out = open(outfile, "w")
+        out = open(outfile, "w", encoding='utf-8')
     ##
 
     # can be 30% faster to write to a tmp file than to pipe to gnuplot
     # bonus, we can keep the plot file for debug.
-    tmp_file = tempfile.NamedTemporaryFile(mode='w',
-                                           suffix='.plt', delete=False)
+    if sys.version_info[0] == 2:
+        tmp_file = tempfile.NamedTemporaryFile(mode='w',
+                                               suffix='.plt', delete=False)
+    else:
+        tmp_file = tempfile.NamedTemporaryFile(mode='w', encoding='utf-8',
+                                               suffix='.plt', delete=False)
     # note that tmp_file is a file handle, it is not a file object
     tmp_file.write(template)
     tmp_file.close()
@@ -1832,7 +1842,7 @@
     header = os.path.join(args.outdir, "header")
     if os.path.isfile(header):
         try:
-            header_file = open(header, 'r')
+            header_file = open(header, 'r', encoding='utf-8')
             header_txt = header_file.read()
             index_buffer += '<br>\n' + header_txt + '\n'
         except IOError:
@@ -1926,7 +1936,7 @@
     footer = os.path.join(args.outdir, "footer")
     if os.path.isfile(footer):
         try:
-            footer_file = open(footer, 'r')
+            footer_file = open(footer, 'r', encoding='utf-8')
             footer_txt = footer_file.read()
             index_buffer += '<br>\n' + footer_txt + '\n'
         except IOError:
@@ -1935,12 +1945,12 @@
 
     # and send the file buffer
     index_filename = os.path.join(args.outdir, "index.html")
-    with open(index_filename + ".tmp", "w") as ifile:
+    with open(index_filename + ".tmp", "w", encoding='utf-8') as ifile:
         ifile.write(index_buffer)
 
     # create csv file, as a tmp file
     csv_filename = os.path.join(args.outdir, "summary.csv")
-    with open(csv_filename + ".tmp", "w") as csv_file:
+    with open(csv_filename + ".tmp", "w", encoding='utf-8') as csv_file:
         csv_ob = csv.writer(csv_file)
         csv_ob.writerow(VizStats.csv_head)
         for row in csvs:
